src/Avalonia.Controls/Utils/SharedSizeScopeHost.cs

```diff
@@ -0,0 +1,400 @@
+
+            public Grid Grid { get; }
+            public MeasurementState MeasurementState { get; private set; }
+
+            public List<MeasurementResult> Results { get; private set; }
+        }
+
+        private readonly AvaloniaList<MeasurementCache> _measurementCaches;
+
+        private class MeasurementResult
+        {
+            public MeasurementResult(DefinitionBase definition)
+            {
+                Definition = definition;
+                MeasuredResult = double.NaN;
+            }
+
+            public DefinitionBase Definition { get; }
+            public double MeasuredResult { get; set; }
+        }
+
+        private class Group
+        {
+            public bool IsFixed { get; set; }
+
+            public List<MeasurementResult> Results { get; } = new List<MeasurementResult>();
+
+            public double CalculatedLength { get; }
+        }
+
+        private readonly Dictionary<string, Group> _groups = new Dictionary<string, Group>();
+
+
+        public SharedSizeScopeHost(Control scope)
+        {
+            _measurementCaches = GetParticipatingGrids(scope);
+
+            foreach (var cache in _measurementCaches)
+            {
+                cache.Grid.InvalidateMeasure();
+                AddGridToScopes(cache);
+
+                cache.GroupChanged.Subscribe(SharedGroupChanged);
+            }
+        }
+
+        void SharedGroupChanged((string oldName, string newName, MeasurementResult result) change)
+        {
+            RemoveFromGroup(change.oldName, change.result);
+            AddToGroup(change.newName, change.result);
+        }
+
+        internal void InvalidateMeasure(Grid grid)
+        {
+            var cache = _measurementCaches.FirstOrDefault(mc => ReferenceEquals(mc.Grid, grid));
+            Debug.Assert(cache != null);
+
+            cache.InvalidateMeasure();
+        }
+
+        internal void UpdateMeasureStatus(Grid grid, GridLayout.MeasureResult rowResult, GridLayout.MeasureResult columnResult)
+        {
+            var cache = _measurementCaches.FirstOrDefault(mc => ReferenceEquals(mc.Grid, grid));
+            Debug.Assert(cache != null);
+
+            cache.UpdateMeasureResult(rowResult, columnResult);
+        }
+
+        private double Gather(IEnumerable<MeasurementResult> measurements)
+        {
+            var result = 0.0d;
+
+            bool onlyFixed = false;
+
+            foreach (var measurement in measurements)
+            {
+                if (measurement.Definition is ColumnDefinition column)
+                {
+                    if (!onlyFixed && column.Width.IsAbsolute)
+                    {
+                        onlyFixed = true;
+                        result = measurement.MeasuredResult;
+                    }
+                    else if (onlyFixed == column.Width.IsAbsolute)
+                        result = Math.Max(result, measurement.MeasuredResult);
+
+                    result = Math.Max(result, column.MinWidth);
```

Comment:
This mirrors what I've seen WPF do - might have some other quirks I'm not aware of.